output/cloud/expv2/metrics_client.go

```diff
@@ -0,0 +1,127 @@
+package expv2
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/golang/snappy"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+
+	"go.k6.io/k6/cloudapi"
+	"go.k6.io/k6/lib/consts"
+	"go.k6.io/k6/output/cloud/expv2/pbcloud"
+)
+
+type httpDoer interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// MetricsClient is a Protobuf over HTTP client for sending
+// the collected metrics from the Cloud output
+// to the remote service.
+type MetricsClient struct {
+	httpClient httpDoer
+	logger     logrus.FieldLogger
+	token      string
+	userAgent  string
+
+	pushBufferPool sync.Pool
+	baseURL        string
+}
+
+// NewMetricsClient creates and initializes a new MetricsClient.
+func NewMetricsClient(logger logrus.FieldLogger, host string, token string) (*MetricsClient, error) {
+	if host == "" {
+		return nil, errors.New("host is required")
+	}
+	if token == "" {
+		return nil, errors.New("token is required")
+	}
+	return &MetricsClient{
+		httpClient: &http.Client{Timeout: 5 * time.Second},
+		logger:     logger,
+		baseURL:    host + "/v2/metrics/",
+		token:      token,
+		userAgent:  "k6cloud/v" + consts.Version,
+		pushBufferPool: sync.Pool{
+			New: func() interface{} {
+				return &bytes.Buffer{}
+			},
+		},
+	}, nil
+}
+
+// Push pushes the provided metrics the given test run.
+func (mc *MetricsClient) Push(ctx context.Context, referenceID string, samples *pbcloud.MetricSet) error {
+	if referenceID == "" {
+		return errors.New("TestRunID of the test is required")
+	}
+	start := time.Now()
+
+	b, err := newRequestBody(samples)
+	if err != nil {
+		return err
+	}
+
+	buf, _ := mc.pushBufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer mc.pushBufferPool.Put(buf)
+
+	_, err = buf.Write(b)
+	if err != nil {
+		return err
+	}
```

Comment:
Unless we are writing directly to the `buf` this is pointless and actually harmful. 

You already created the `[]byte` slice and now you are going to put it in a pool so it will survive longer. Only to get it out and not use it but just write on top of it.

If we are not going to be writing to it (seem my comments on `newRequestBody`) it is better to just use `bytes.Reader` and skip the pool.

Reply:
You're right, this is some remaining stuff from the old client. Done.